pkg/webhook/pod.go

```diff
@@ -0,0 +1,197 @@
+
+// PodMutator implements controller-runtime WebHook interface.
+type PodMutator struct {
+	decoder  *admission.Decoder
+	cfg      *Config
+	Mutators []MutatorConfig
+}
+
+type MutatorConfig struct {
+	Mutator  Mutator
+	Required bool
+}
+
+type Mutator interface {
+	ID() string
+	Mutate(ctx context.Context, p *corev1.Pod) (newP *corev1.Pod, changed bool, err error)
+}
+
+func (pm *PodMutator) InjectClient(_ client.Client) error {
+	return nil
+}
+
+// InjectDecoder injects the decoder into a mutatingHandler.
+func (pm *PodMutator) InjectDecoder(d *admission.Decoder) error {
+	pm.decoder = d
+	return nil
+}
+
+func (pm *PodMutator) Handle(ctx context.Context, request admission.Request) admission.Response {
+	// Get the object in the request
+	obj := &corev1.Pod{}
+	err := pm.decoder.Decode(request, obj)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	newObj, changed, err := pm.Mutate(ctx, obj)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	if changed {
+		marshalled, err := json.Marshal(newObj)
+		if err != nil {
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+
+		// Create the patch
+		return admission.PatchResponseFromRaw(request.Object.Raw, marshalled)
+	}
+
+	return admission.Allowed("No changes")
+}
+
+func (pm PodMutator) Mutate(ctx context.Context, p *corev1.Pod) (newP *corev1.Pod, changed bool, err error) {
+	newP = p
+	for _, m := range pm.Mutators {
+		tempP := newP
+		tempChanged := false
+		tempP, tempChanged, err = m.Mutator.Mutate(ctx, tempP)
+		if err != nil {
+			if m.Required {
+				err = fmt.Errorf("failed to mutate using [%v]. Since it's a required mutator, failing early. Error: %v", m.Mutator.ID(), err)
+				logger.Info(ctx, err)
+				return p, false, err
+			} else {
+				logger.Infof(ctx, "Failed to mutate using [%v]. Since it's not a required mutator, skipping. Error: %v", m.Mutator.ID(), err)
+				continue
+			}
+		}
+
+		newP = tempP
+		if tempChanged {
+			changed = true
+		}
+	}
+
+	return newP, changed, nil
+}
+
+func (pm *PodMutator) Register(ctx context.Context, mgr manager.Manager) error {
+	wh := &admission.Webhook{
+		Handler: pm,
+	}
+
+	mutatePath := getPodMutatePath()
+	logger.Infof(ctx, "Registering path [%v]", mutatePath)
+	mgr.GetWebhookServer().Register(mutatePath, wh)
+	return nil
+}
+
+func (pm PodMutator) GetMutatePath() string {
+	return getPodMutatePath()
+}
+
+func getPodMutatePath() string {
+	pod := flytek8s.BuildIdentityPod()
+	return generateMutatePath(pod.GroupVersionKind())
+}
+
+func generateMutatePath(gvk schema.GroupVersionKind) string {
+	return "/mutate-" + strings.Replace(gvk.Group, ".", "-", -1) + "-" +
+		gvk.Version + "-" + strings.ToLower(gvk.Kind)
+}
+
+func (pm PodMutator) CreateMutationWebhookConfiguration(namespace string) (*admissionregistrationv1.MutatingWebhookConfiguration, error) {
+	caBuff, err := ReadFile(filepath.Join(pm.cfg.CertDir, "ca.crt"))
```

Comment:
why cant we use ioutil.ReadFile? buffer?

Reply:
yeah... I was using it elsewhere that needed a buffer... but not anymore... good call!